x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -347,6 +348,60 @@ public void testRunAsAuthenticationWithDomain() {
         assertThat(test.getDomain(), nullValue());
     }
 
+    public void testCanAccessAcrossDomain() {
+        RealmRef authRealmRef = randomRealmRef(randomBoolean());
+        User authUser = randomUser();
+        Authentication realmAuthentication = Authentication.newRealmAuthentication(authUser, authRealmRef);
+
+        RealmDomain otherDomain = randomDomain(randomBoolean());
+        RealmConfig.RealmIdentifier otherRealmIdentifier = randomFrom(otherDomain.realms());
+        RealmRef otherRealmRef = new RealmRef(
+            otherRealmIdentifier.getName(),
+            otherRealmIdentifier.getType(),
+            randomAlphaOfLengthBetween(3, 8),
+            otherDomain
+        );
+
+        // same user
+        Authentication otherRealmAuthentication = Authentication.newRealmAuthentication(authUser, otherRealmRef);
+        if ((authRealmRef.getDomain().realms().contains(otherRealmIdentifier))
```

Comment:
> It seems to me that this can throw NPE because authRealmRef.getDomain() can be null?

That's correct I fixed it.

> Also it is very unlikely for this check to be true because realm names are randomized. So it is unlikely to have two random realms to pass the equals check. So the chance of this branch being executed for two realms that are not file or native is very low. It might be OK since tehnically it is still tested but just with a very low possibility? Or do we need do something to make the test condition happen more often?

Also correct, but the goal was not to test the access across a domain accessing file and native realms.
I thought about it, and it didn't sound significant enough to warant a dedicated test in the "domain" scope.
There's a bit too much randomness in these tests overall, and that is a problem.
The root of it is that there are many species of Authentications.